File: python/aoc/puzzle_4.py
#!/usr/bin/env python

import sys


def main():
    aim = 0
    depth = 0
    horizontal = 0

    for line in sys.stdin.readlines():
        action, amount = line.split()
        amount = int(amount)

        if action == "forward":
            horizontal += amount
            depth += aim * amount
        elif action == "down":
            aim += amount
        elif action == "up":
            aim -= amount
        else:
            raise NotImplementedError(action)

    print(depth * horizontal)


if __name__ == "__main__":
    main()

File: rust/puzzle_4.rs
use std::{io, io::prelude::*};

fn main() {
    let mut aim: usize = 0;
    let mut depth: usize = 0;
    let mut horizontal: usize = 0;

    for line in io::stdin().lock().lines() {
        let l = line.unwrap();
        let (action, amount) = l.split_once(" ").unwrap();
        let amount: usize = amount.parse().unwrap();

        match action {
            "forward" => {
                horizontal += amount;
                depth += aim * amount;
            }
            "down" => aim += amount,
            "up" => aim -= amount,
            _ => (),
        };
    }

    println!("{}", depth * horizontal);
}
